Paragraph: Add LineHeightControl + Attribute

ItsJonQ opened the pull request:
## Description
<img width="1042" alt="Screen Shot 2020-03-10 at 13 34 59" src="https://user-images.githubusercontent.com/2322354/76341847-fbc78580-62d3-11ea-97c4-b3e167dfb06d.png">

This update adds a new `lineHeight` attribute to the core Paragraph block. Accompanying it would be a new dedicated (and experimental) `LineHeightControl`, which is provided by `@wordpress/block-editor`.

## How has this been tested?

This has been tested on local Gutenberg on multiple themes, from 2011 -> 2020.

## Screenshots 

![Screen Capture on 2020-03-10 at 13-32-25](https://user-images.githubusercontent.com/2322354/76342014-3d583080-62d4-11ea-86f7-31334a798132.gif)

GIF demonstrates the live control edit -> rendering for line heights.

## Types of changes

A majority of the updates in this PR hopes to improve the relationship + setup of attributes and controls.
A goal for `LineHeightControl` is to be implemented with **minimal code changes** for a Block's Edit/Save components.

### `LineHeightControl`

Taking a look at the `Paragraph` `edit.js`, the new `LineHeightControl` has been added under `InspectorControl` without requiring any props.

The component is able to retrieve the appropriate `attributes` and `setAttributes` on it's own. This workflow makes it much simpler to add controls to Edit components - especially if the control requires multiple moving parts.

### `useBlockEditProps()`

This new experimental hook from `@wordpress/block-editor` . This magical hook (which leverages other new experimental hooks) is able to retrieve a block's `attributes` and `setAttributes` without requiring anything to be passed in.

The new `LineHeightControl` relies on this hook as the getter/setter mechanism for the `lineHeight` value.

If folks are 👍 with this approach, I can see it being extremely useful for attributes x controls moving forward 🙌 

### Unitless

To keep things simple, the `line-height` value will be unitless and works as a multiplier to the Paragraph's font-size. This more hands-off approach allows for greater flexibility for font-size x line-height.

### Control Styles

Some `@emotion` packages were added as a dependency to `@wordpress/block-editor` to enable CSS-in-JS style rendering for `LineHeightControl`. These are the same packages used internally within `@wordpress/components`.

### UX Interactions

To improve the UX for `LineHeightControls`, I implemented some logic to better handle incrementing/decrementing values from the start.

For example...

If you're adjusting `line-height` for the **first time**, and you press `UP`, the value would be `1.6`, rather than the default `input` behaviour of `0.1`.

Alternatively, if you press `DOWN` for the **first time**, the value would be `1.4`, rather than `0`.

This is because it uses a smart default of `1.5`. That being said, `1.5` isn't actually applied. There is no `line-height` style set **until** the user makes a change.

Afterwards, incrementing/decrementing works as normal.
Removing the value would reset it entirely.

During development, I found that the above behaviour adjustment was key. Without it, the initial increment (`UP` press), would cause all of the text to collapse, as it would use a `line-height` value of `0.1`.

### Tests

I would love to write tests for this new control. I attempted to do this, but I wasn't able to successfully mock the new `useBlockEditProps` hook, which technically could be replaced with `useState`. Jest kept complaining 😅 

## Closing Thoughts

Lots of changes in this PR! I'm open to anything. On the surface, adding a new `line-height `property to Paragraph is a very minor thing. The bigger goal though would be to systematize and make it easier for blocks (both core and 3rd party) to set up attributes and render their respective controls.

Thank you! 🙏 

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Partially takes care of:
https://github.com/WordPress/gutenberg/issues/20339

Commit:
Update emotion library to latest. Adds to block-library package.

Commit:
Add LineHeightControl component to block-editor

Commit:
Export LineHeightControl as experimental from block-editor

Commit:
Hook up LineHeightControl with core Paragraph block

Commit:
Improve undefined/reset interactions for line-height attribute

Commit:
Refactor store hook usage with LineHeightControl

Commit:
Refactor with withLineHeight HOC for Edit and Save components

Commit:
Improve reset logic. Simplifies stlye value output.

Commit:
Refactor function names and comments

Commit:
Fix import/export of new experimental attribute hooks

ItsJonQ commented on packages/block-library/src/paragraph/block.json:
```diff
@@ -33,12 +33,18 @@
 			"type": "string",
-			"enum": [ "ltr", "rtl" ]
+			"enum": [
+				"ltr",
```
Prettier autofixed these ones

ZebulanStanphill commented:
I am very much against this going in as a control for individual blocks. In my opinion, this is pure bloat; an option that gives too much flexibility and tries to solve a problem the wrong way. Any line height option should be handled at the global styles level; and I don't mean changing this for every Paragraph block... I mean this should be a setting at the same level as a site wide typography scale.

There is no good reason to add this kind of control to individual blocks like this. If a user is needing to change the line height on a single block, then that's a sign that the global styles were not good enough to begin with. The problem should be solved at the source, not with a band-aid like this. Please, do not add this control to the Paragraph block (or any block, really).

ZebulanStanphill commented:
I should note that I'm fine with adding a component for this that could be used in global styles; I'm just highly against this kind of control being provided at the individual block level.

mtias commented:
I disagree — most of the patterns being designed cannot be accomplished without control over line-height at the block level. If there is control over font-size, whether exposed to users or not, there should be control over line-height as they go hand in hand.

![image](https://user-images.githubusercontent.com/548849/76347586-2ed73c80-6307-11ea-8cca-8f825945e0c6.png)

This is entirely complimentary to global styles handling a typography scale, the theme defining its own defaults, and so on. The _block itself_ needs to understand the attribute, so this is a step in that direction.

It should be made easy to disable globally like other customization tools, of course, and we could debate the argument of whether it should be exposed by default or implicit on any defined pattern.

ZebulanStanphill commented:
@mtias Why not just make line heights part of the font size presets? We should definitely be encouraging the use of the font size presets rather than custom pixel sizes, and if we already have presets, why not automatically set the appropriate line height?

(These font size + line height combos could be defined at the theme level or at the global styles level.)

We already add padding when selecting a background color on various blocks, so the precedent for such a thing is already there. The user shouldn't have to be responsible for selecting the right line height to go with a font size. They shouldn't even have to think that much about the font size they want to use. They should be able to pick from a few fine-tuned presets and have it work by default with no extra fiddling around.

mtias commented:
I think we are mostly talking about the same thing, in a way :) For visual effect, there will still be cases where you need specific control over it.

![image](https://user-images.githubusercontent.com/548849/76354712-44eafa00-6313-11ea-9143-87b680fb8751.png)

This doesn't mean that we cannot absorb most common cases in pairs (this we should do regardless for the main font-sizes) and global styles / theme, as you say, but it does mean we need the concept of its control at the block level somehow existing too.

ZebulanStanphill commented:
@mtias I think I get the idea of that example, but even in that situation, such an intentional visual effect is likely to be used in multiple places, and so something like that example would be far more suited to a block style variation for the Paragraph block (and/or Heading block).

Intentional stylistic choices like that should be the responsibility of the website designer, and good design reuses parts to create a cohesive whole. Therefore, style variations seem like the proper way to implement this.

By adding a line-height control to individual Paragraph blocks, you encourage one-off modifications. You bring to the forefront a "solution" to a problem, but it's the wrong solution in nearly every situation. You give the content creator more power, but also more responsibility... responsibility that what supposed to be handled by the site designer (whether that be the theme author or the person managing the global styles or both).

Individual block options should be designed around carefully crafted presets set by the designer, thus keeping the responsibility of design with the designer, not the content editor.

I recently read [a very interesting blog post](https://thinkwritecode.com/the-good-wordpress-theme/) that points out how Gutenberg seems to be blurring the lines of responsibility with its increasing amount of style options on individual blocks.

But it doesn't have to be that way. We already have the block style variations API, and global styles are going to become a thing. Only weird one-off cases are not handled, and one-off cases should not be encouraged, in my opinion.

I think a better solution to this problem would be to provide an easier way to create/register a style variation. We could provide a visual interface for creating style variations so people don't have to write CSS; in that UI, there can be a line height control, as well as a zillion other advanced style options. But in the end, that style variation will be just a class applied to a block. There's no need to add a line height attribute to any block.

ZebulanStanphill commented:
Additionally, it follows based on my reasoning that the custom font size option probably shouldn't even exist for individual blocks. That may be a controversial stance, and it's definitely useful right now since global styles are not yet implemented and people can't easily create style variations, but in the long term, the custom font size option is not something that should be exposed for individual blocks. In the long run, custom font sizes should be opt-in, not opt-out. As it is, themes have lost control over font size consistency.

At least in the case of font sizes, the custom option is somewhat buried, with the preset options presented first. The same can not be said of this line height control.

chrisvanpatten commented:
I know this is probably going to come with the global styles work, but it'd be neat to see a "make default" option, so if you change the setting you can persist it back to the global configuration.

I do strongly feel like this would be a good candidate for the features API proposed in #20623, or at the very least something with the theme supports we currently have (`add_theme_support( 'disable-custom-line-height' )` to mirror `add_theme_support( 'disable-custom-font-sizes' )` and such).

ZebulanStanphill commented:
I agree that at the very least, themes should be able to disable this feature similar to custom font sizes and custom colors.

Also, another thing I wanted to point out: in the aforementioned stylistic flair example posted by @mtias, the person wanting to do that could just as easily have wanted to also use a different font weight or letter spacing. Should we be adding font weight and letter spacing controls to the Paragraph block? I don't think so.

mtias commented:
> But in the end, that style variation will be just a class applied to a block. There's no need to add a line height attribute to any block.

@ZebulanStanphill this is not necessarily the case. The actual appearance attributes need to make it to the block as _attributes_, otherwise we cannot ensure true WYSIWYG. Also mobile apps don't understand class names with arbitrary CSS attached, they will need to consume agreed upon attributes. The way global styles would work is by traversing the actual appearance attributes within the editor so that they can be consumed at the block level as well. This is paving the way for that, if seen beyond the mere presence of the line-height control on the block, as it makes the block understand an attribute. The presence of a control by the block is just an artefact. We also need to see how all of these should eventually be grouped in a style panel on the block, including ways to make changes the default (or save style variations) but all these layers require we build the tools, otherwise we are chasing ghosts.

> By adding a line-height control to individual Paragraph blocks, you encourage one-off modifications. You bring to the forefront a "solution" to a problem, but it's the wrong solution in nearly every situation. You give the content creator more power, but also more responsibility... responsibility that what supposed to be handled by the site designer (whether that be the theme author or the person managing the global styles or both).

There's a multitude of use cases (one of pages, brochures, eclectic presentations) and we should not force our own preference on what site design should be on all people. Most block libraries have added these controls, but the implementation ends up being generally done through inline styles which hurts the user in the longer term. By taking on that mantle we can ensure consistency and build the right levels of control, which is one of the main reasons to do this in core, to avoid fragmentation and ensure the user stands on solid ground.

Site builders will be able to retain full control by not enabling / disabling certain appearance attributes. Once it's all properly codified, that should be an easy configuration set globally and cascading properly to blocks because they understand what to do with the attributes. This is also true if we want global styles to transpire to third-party blocks.

It's important to separate the fact getting the controls in place doesn't mean that everything will be exposed to end users. Most of the design tools, including global styles, theme editor, etc, are likely going to be admin level to begin with. The reality remains that global styles cannot work without also building the tools, and starting on individual blocks is more ergonomic than trying to solve everything at once at the global level. It also happens that patterns need more fine grained control over these attributes.

> but it'd be neat to see a "make default" option, so if you change the setting you can persist it back to the global configuration.

@chrisvanpatten agreed, this was in one of the original block style variations issues as one of the future tasks.

Also, let's move the main conversation to the issue: #20339 !

chrisvanpatten commented:
One thing I want to observe is in regards to `useBlockEditProps()`. This is just an observation, not intended to be firm or anything — I'm still formulating my own thoughts and thinking out loud here.

As noted in the original post above, this is indeed a magical hook! I've been wondering for a while if this type of pattern might crop up.

I think it's a pattern that clearly has its use-cases — I fully switched our codebase from `withState` to `useState` because of how much I like hooks — but I'm also wondering about the potential impact on clarity and readability. One thing I really like about the `XyzControl` pattern in Gutenberg is the simplicity, consistency, and flexibility of the `onChange` and `value` props. It's easy for a beginner to grok and makes the code a little more clear.

As awesome as hooks are, they do mask away some explicitness in favor of magic. And as great as magic can be, I think it's important to balance alongside readability and clarity.

Again, this isn't to say that we should go one way or the other… just wanted to share some musings that I've had recently that this hook brought to mind in particular.

I think technically the work is super interesting here, and seems like a great place to trial some exciting new APIs!

mtias commented:
> As awesome as hooks are, they do mask away some explicitness in favor of magic. And as great as magic can be, I think it's important to balance alongside readability and clarity.

Yes, this is a tricky thing to balance. Explicitness is sound when it comes to the main attributes of a block, but when we start dealing with shared attributes, it becomes unnecessarily verbose and tedious to add support. Adding support for text and background color means adding four new attributes at the moment! That's why I asked to explore the notion of implicit attributes as well: https://github.com/WordPress/gutenberg/pull/20308

karmatosed commented:
First, I am super excited to see this. There is a strong case for line-height to create designs and really adjust to make the perfect layout that people envision in their head.

@ZebulanStanphill whilst I understand your point of view on this, I would ask you to consider a few points. If we are to truly give control, I don't think a range hits the right spot. There are many cases where more granular control really is what is needed. For sure we should offer strong defaults, that is where ideas like having a range control for typography scale come in. 

We do need to offer some tools we don't right now to empower people to create. This doesn't always work per block, sometimes we're going to need to do globally, sometimes per block. However, we do need to open up the tools we offer to provide a powerful kit to create what people really want. It's exciting if we do that, because there are going to be so many opportunities for incredible things to be made.

Commit:
Export LineHeightControls and withLineHeight for native

Commit:
Adjust mechanism to get block attributes in useBlockAttributes hook

karmatosed reviewed (approved):
Testing this and giving a design review, it's great. I am not giving a code review so that needs to happen but this is exactly as planned so really thank you.

Commit:
Add theme support for disable-custom-line-height

ZebulanStanphill commented on lib/client-assets.php:
```diff
@@ -659,3 +659,17 @@ function gutenberg_extend_settings_block_patterns( $settings ) {
+ * @return array Filtered editor settings.
+ */
+function gutenberg_extend_settings_custom_line_height( $settings ) {
+	$settings['disableCustomLineHeight'] = get_theme_support( 'disable-custom-line-height' );
```
Thank you! :+1:

ZebulanStanphill commented on packages/block-editor/src/components/line-height-control/utils.js:
```diff
@@ -0,0 +1,99 @@
+import { __experimentalUseBlockEditProps as useBlockEditProps } from '../../store';
+
+export const BASE_DEFAULT_VALUE = 1.5;
+export const INITIAL_VALUE = '';
```
It seems weird to have `INITIAL_VALUE` be a string when the attribute is supposed to be an integer. Perhaps setting it to `null` would make more sense?

ItsJonQ replied:
@ZebulanStanphill That's a great observation! I just pushed some refactors and added comments to clarify it a bit:

```
/**
 * There are varying value types within LineHeightControl:
 *
 * {undefined} Initial value. No changes from the user.
 * {string} Input value. Value consumed/outputted by the input. Empty would be ''.
 * {number} Block attribute type. Input value needs to be converted for attribute setting.
 *
 * Note: If the value is undefined, the input requires it to be an empty string ('')
 * in order to be considered "controlled" by props (rather than internal state).
 */
```

Commit:
Clarify logic + add comments for lineHeight attribute -> value evaluation

Commit:
Simplify CSS output for lineHeight in paragraph block

Commit:
Add experimental prefix for disableCustomLineHeight setting

ItsJonQ commented:
Thank you all for the discussion and feedback! 

> As noted in the original post above, this is indeed a magical hook! I've been wondering for a while if this type of pattern might crop up.

@chrisvanpatten  😊 . I use the term magic in a playful manner. It's more of a simplification vs. hackery (as typical magic goes) ✌️ . Under the hood, it uses existing hooks provided by `@wordpress/data`. It just combines various bits together into a single easy to use function :).

Chris + @ZebulanStanphill Thank you for your thoughts on `add_theme_support`! I've added that functionality in there.

ZebulanStanphill commented on packages/block-library/src/paragraph/style.scss:
```diff
@@ -28,6 +28,10 @@
 	font-style: normal;
 }
 
+:root p.has-line-height {
```
Why is the `:root` necessary?

ItsJonQ replied:
It increases the specificity of the style for more consistent rendering in the editor and on the front-end (considering typical theme styles)

ZebulanStanphill replied:
Alright, then. Perhaps a comment may be helpful?

paaljoachim commented:
Just a short question...
After this has been merged are there also plans to add a similar control to change the text width? It could be be added to the empty space right next to the line height. 
Btw Thanks for working on this @ItsJonQ

iamtakashi commented:
Thank you for working on this. I can't wait for the line-height control to land! This has been one of the big missing design tools blocking for me to build block patterns with a custom font size.

karmatosed commented:
> After this has been merged are there also plans to add a similar control to change the text width?

Do you mean the block width or font style? There is an issue for font styling and width of a block. We have to balance how many controls so just trying to find out what you mean by font width @paaljoachim

paaljoachim commented:
Hi Tammie.
I appologize. I was unclear.
I am thinking letter spacing: https://github.com/WordPress/gutenberg/issues/20796
It would fit well in beside the line height control.

youknowriad commented on packages/block-editor/package.json:
```diff
@@ -23,6 +23,8 @@
 	"dependencies": {
 		"@babel/runtime": "^7.8.3",
+		"@emotion/core": "^10.0.28",
+		"@emotion/styled": "^10.0.27",
```
Should we just keep the CSS in JS in the components package for now for consistency?

ItsJonQ replied:
Sure. I can refactor this 👌

youknowriad commented on packages/block-editor/src/components/line-height-control/index.js:
```diff
@@ -0,0 +1,73 @@
+
+export { default as withLineHeight } from './with-line-height';
+
+export default function LineHeightControl( props ) {
```
This is  departure from existing block-editor components where we have value and onChange props. 
I guess there's room for magical components like that that consume directly from attributes but I feel that this is not the right PR to introduce these:

 - These make strong assumptions about  the naming of the attributes
 - If we were to introduce the implicitness involved here, I'd personally prefer if it's a dedicated PR with a plan to generlize this to all block editor components instead of an adhoc one.

ItsJonQ replied:
> I guess there's room for magical components like that that consume directly from attributes but I feel that this is not the right PR to introduce these:

@youknowriad Thank you for your feedback!
Open question! When would it be the right time to introduce this pattern?

I think simplifying Control based components would help simplify the system. Other candidates would be existing color and font size attributes.

> If we were to introduce the implicitness involved here, I'd personally prefer if it's a dedicated PR with a plan to generlize this to all block editor components instead of an adhoc one.

I thought it being behind an `__experimental` prefix would have been okay.
I viewed this as more a first attempt at this approach since it would only be scoped to the Paragraph block.

youknowriad replied:
> I viewed this as more a first attempt at this approach since it would only be scoped to the Paragraph block.

I think that's a good way to introduce it iteratively. I know @mtias likes this pattern but I have concerns personally though. The main one being the implicit attribute presence. 

One way to lower that issue though would be to use a generic attribute. For example `style` could work as it can be an attribute used for any global style variables. 

If we use separate attributes, I'd worry about the proliferation of adhoc attributes.

youknowriad commented on packages/block-editor/src/components/line-height-control/with-line-height.js:
```diff
@@ -0,0 +1,45 @@
+ * Higher-order component that provides a Block's Edit/Save component with
+ * interpolated className and style based on the lineHeight attribute value.
+ */
+export default function withLineHeight() {
```
This is also a "new thing". using HoCs for "save" functions is a new pattern that raises some concern:

 - The implicitness/magic might not be a desired behavior
 - The difference with all the remaining attributes.
 - Lack of flexibility as it assumes that the className and styles across attributes are applied to the same element (not sure if it's always the case)

ItsJonQ replied:
> Lack of flexibility as it assumes that the className and styles across attributes are applied to the same element

I can see this one being a potential concern 🤔 
Do you have any suggestions for the other 2 points?

mcsf replied:
I'd also strive to avoid magic in crucial places such as `save`. 

> Do you have any suggestions for the other 2 points? (With the goal of reducing verbosity in the `save` function)

It's clear that we need #20308 to yield some concept that we can try out regarding implicit attributes. I realise that React hooks currently cannot be used inside `save`, which further constrains us (and that's partly a good thing), otherwise I might say that we could solve this with a hook that would collect any implicit attributes that need to make their way into aspects of the serialised block.

An alternative I wonder about is how much we could leave to the server to deal with this instead, or how far we could take Andrew's idea of leveraging the unit of "the block" as a marker or carrier of style that is composable.

ItsJonQ replied:
> otherwise I might say that we could solve this with a hook...

@mcsf If only save worked with hooks 😊. As you pointed out, I went with the solution I did due to the save function not supporting hooks. The approach I took is similar in spirit with https://github.com/WordPress/gutenberg/pull/20308 and `useColor`, although perhaps not as daring. 

Ultimately, I think we need a systematic way to handle the setting/rendering of these style based attributes. It sounds like that is something we all agree upon 👍 . The question is when, where, and how.

This is something I've been chatting with @mtias about.

If folks are uncomfortable with the experimental abstraction at this strange, I'm happy to pull it back to follow the patterns we currently have.

mcsf replied:
@ItsJonQ, thanks for indulging me.

> If folks are uncomfortable with the experimental abstraction at this strange, I'm happy to pull it back to follow the patterns we currently have.

I don't want this to block your exploration, so I'll leave this up to you, as long as the solution is something well contained that only applies to Paragraph (and hopefully doesn't mislead others in thinking they should follow a new design pattern).

My personal preference might be to, as you say, pull back and make the changes to `save` here as explicit and concrete as possible — leaving any abstractions to a stage at which we not only have better tooling but also broader understanding of what is needed. That said, you can still choose to encapsulate some of the functionality; if so, keep [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) in mind.

ItsJonQ replied:
Thanks for the feedback @mcsf !

I'll pull back the abstraction for now. Just a heads up, we're going to be [adding LineHeight controls for **Heading** as well](https://github.com/WordPress/gutenberg/issues/20339).

Perhaps after we remove the abstraction for Paragraph... We can revisit it (or something like it) when we add it to Heading.

mtias replied:
That makes sense to me.

Commit:
Add comment for :root based CSS rule

Commit:
Removes @emotion from block-library. Refactor to use scss

youknowriad commented on packages/block-library/src/paragraph/style.scss:
```diff
@@ -28,6 +28,12 @@
+// :root increases CSS specificity
+// This improves rendering consistency within the editor and site.
+:root p.has-line-height {
+	line-height: var(--wp--core-paragraph--line-height, inherit);
```
is this in line with the global styles convention? Is this variable supported by the global styles theme.json, should we add a fallback to the block-less variable? cc @jorgefilipecosta

mcsf commented on packages/compose/README.md:
````diff
@@ -188,7 +188,7 @@ Returns true if the viewport matches the given query, or false otherwise.
 
 ```js
-useViewportMatch( 'huge', '<' );
+useViewportMatch( 'huge', <' );
````
Accidental edit?

ZebulanStanphill replied:
This typo was in `master` until #20911, so this is probably a rebase mistake.

Commit:
Remove withLineHeight HOC abstraction.

Reverted to use inline styles instead of CSS variable for line-height style rendering

Commit:
Revert className implementation

Also revert @emotion library versions to master

ItsJonQ commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -109,6 +118,13 @@ function ParagraphBlock( {
 		[ fontSize.size ]
 	);
 
+	const styles = {
```
Created a variable from inline implementation

Commit:
Revert @emotion/native from 10.0.23 -> 10.0.22 (to match master)

ItsJonQ commented:
@youknowriad , @mtias , @mcsf 

👋 Haii!!

I've removed the HOC abstraction and adjusted the rendering to match what we currently have for the other blocks.

Let me know if you feel like anything else needs to be adjusted 🤞

youknowriad commented:
@ItsJonQ Thanks for the update. This is looking better for me.

One thing I'm wondering though is this https://github.com/WordPress/gutenberg/pull/20775#discussion_r393068216

While the current implementation aligns perfectly with existing attributes, I'm wondering if we should use this PR as a starting point to support "Global styles related attributes".

What I mean by "Global styles related attributes" is attributes that generate a CSS variable instead of an explicit inline style.  That same variable could be defined globally as well using the `theme.json`  or globally per block.

Also, related if we follow this path, is whether we'd add a dedicate attribute (like `line-height`) here or a generic `style` attribute from which the variables can be automatically generated using the keys  of that style attribute (object) and a naming convention for these variables. 

Thoughts @jorgefilipecosta @mtias @mcsf

ItsJonQ commented:
@youknowriad It would be awesome to try out CSS variables here. I'm not sure of the exact implementation strategy yet (at this state of Global Styles).

Would love to hear from others!

Commit:
Updates package-lock.json

Commit:
Fix e2e test to target correct text input

aduth commented on package-lock.json:
```diff
@@ -1878,6 +1878,7 @@
 			"version": "10.0.19",
 			"resolved": "https://registry.npmjs.org/@emotion/cache/-/cache-10.0.19.tgz",
 			"integrity": "sha512-BoiLlk4vEsGBg2dAqGSJu0vJl/PgVtCYLBFJaEO8RmQzPugXewQCXZJNXTDFaRlfCs0W+quesayav4fvaif5WQ==",
+			"dev": true,
```
If there are no changes to a `package.json`, it's not clear to me why this file would have been changed?

ItsJonQ replied:
@aduth I'm not sure 🙈 . I think it had to do with my `npm` version. I'll revert this to the one from `master`.

jorgefilipecosta commented on packages/block-editor/src/store/hooks.js:
```diff
@@ -0,0 +1,65 @@
+ * @return {Object} Block attributes.
+ */
+export function useBlockAttributes() {
+	const clientId = useSelectedBlockClientId();
```
While this pattern using hooks inside other hooks is nice and makes the code very readable, I think it has some performance cost. For example, on each useBlockAttributes call we are going to subscribe to the store in useBlockAttributes and in useSelectedBlockClientId while if we did not use useSelectedBlockClientId in useBlockAttributes and just called the selector getSelectedBlockClientId in useBlockAttributes we would have only one subscription.
Is this something we should worry about or the advantages of having code more readable worth the addition store subscriptions? cc: @aduth, @youknowriad

jorgefilipecosta commented on packages/block-editor/src/store/hooks.js:
```diff
@@ -0,0 +1,65 @@
+	const attributes = useSelect( ( select ) => {
+		const { getBlockAttributes } = select( 'core/block-editor' );
+		return getBlockAttributes( clientId ) || {};
+	}, [] );
```
We should include clientId in the dependencies array.

Commit:
Reverting package-lock.json

ItsJonQ commented on packages/e2e-tests/specs/editor/various/embedding.test.js:
```diff
@@ -307,7 +307,7 @@ describe( 'Embedding content', () => {
 		await publishPost();
 		const postUrl = await page.$eval(
-			'[id^=inspector-text-control-]',
+			'.editor-post-publish-panel [id^=inspector-text-control-]',
```
That's the interesting thing! For the longest time, this test was failing for this PR.

The reason for that is because the `Paragraph` block adds a `TextControl` component.
When this test `$eval` for an input, it finds the new LineHeightControl, **not** the publish URL input.

To resolve this, I made the test more specific to the Publish Panel

jorgefilipecosta replied:
Thank you for clarification @ItsJonQ!

mtias commented on packages/block-library/src/paragraph/block.json:
```diff
@@ -33,12 +33,18 @@
 		"fontSize": {
 			"type": "string"
 		},
+		"lineHeight": {
```
Perhaps, but not as part of this PR.

jorgefilipecosta replied:
@mtias, so should we merge this PR with the new attribute and then when we refactor to use style attribute look into a migration for the attribute?

mtias replied:
Hopefully we can figure out between merging this and releasing the next plugin update so that we don't need to deprecate / migrate.

youknowriad replied:
yes, the refactor should be easy enough once we get #21021 (which is almost ready, just need to fix the docgen tool to support optional chaining).

Let's wait after the RC though.

mtias replied:
Agreed.

Commit:
Provide clientId in dependency array in useSetAttributes hook

mtias commented on packages/block-library/src/paragraph/save.js:
```diff
@@ -47,6 +49,7 @@ export default function save( { attributes } ) {
 		backgroundColor: backgroundClass ? undefined : customBackgroundColor,
 		color: textClass ? undefined : customTextColor,
 		fontSize: fontSizeClass ? undefined : customFontSize,
+		lineHeight: isUndefined( lineHeight ) ? undefined : lineHeight,
```
I think we should, yes, but I'd prefer if we do it with a couple attributes to see how it'd work more clearly.

jorgefilipecosta commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -131,6 +147,7 @@ function ParagraphBlock( {
 						value={ fontSize.size }
 						onChange={ setFontSize }
 					/>
+					<LineHeightControl />
```
The way to use the control is nice. One does not need to pass anything, but it is different from the way we use all the other controls. I would expect that in a control component, I would be able to pass a value, and onChange so I could store the data in the way I want or do something when the value is changed.
Would it make sense to support both usages e.g.: we use attributes and the way things work right now when value and onChange are not passed, but we still allow value and onChange?
I guess as a follow-up, we may see if other controls we have right now could also include support for directly using attributes.

mtias replied:
As part of having these implicit attributes, I'd love to see if we can reduce the need to handle attribute changes.

ItsJonQ replied:
Interesting point! Maybe the naming of the component would be different as it's API is different compared with our other controls.

The idea for the new `LineHeightControl` is to make it easy to plop in. As you noted, it's nice how simple it is 😊 .

We technically could support both use cases. Although... I'm not sure if we need (thinking out loud). The control itself doesn't really do much. It just connects the set + setState mechanisms with a `TextControl` primitive.

If a user needed to do this, they could replicate the same thing.
For more complex controls (e.g. the [UnitControl](https://github.com/WordPress/gutenberg/pull/20888) I'm adding for Cover), we can create a special primitive as part of `@wordpress/components`, and have the hooked up component within `@wordpress/block-editor`.

One primitive for customized implementation.
One preconfigured for a prop-free integration

chrisvanpatten replied:
I agree a new naming convention would make sense here. It also goes back to my earlier concerns where I would expect a `*Control` to have the `onChange` and `value` props, and was concerned about some of the magic. The naming convention has a lot to do with that expectation.

I don't have a better idea for the naming convention but I do think that's a big part of it.

ItsJonQ replied:
> The naming convention has a lot to do with that expectation.

@chrisvanpatten That's a fair point. In that case... what do folks think about:

If *no props passed*... `<LineHeightControl />` would implicitly use values from `attributes`.

Otherwise, if `value` and `onChange` props are passed, `<LineHeightControl />` would use those instead.

jorgefilipecosta replied:
> If no props passed... <LineHeightControl /> would implicitly use values from attributes.

> Otherwise, if value and onChange props are passed, <LineHeightControl /> would use those instead.

This seems like a good approach 👍

youknowriad replied:
For me and since we're going to try an approach like #21021 where block authors won't have to use that component directly and would just enable a flag, I feel the magic/implicit approach won't be needed anymore. The implicitness will come from the hooks, not from the component itself.

I might take some time today to try to see how to apply the colors approach to the line-height feature here.

ItsJonQ replied:
Awesome! Interested to see how that goes! 🙏

youknowriad replied:
I implemented this in 6c7dcbaa94 

Here are my takeaways:

 - I like the consistency with other styles (color for now explored in #21037) and the simplicity.  
 - It's neat to be able to define support for a customization with a single line `__experimentalLineHeight: true`
 - It's cool that it's based on CSS variables for Global styles support, this has the downside of potentially requiring some CSS tweak from theme authors in order to have less specificity for their line-height styles to avoid overriding the CSS variable. (I think this work is needed regardless of this PR)
 - Right now, the biggest downside is that we end up with two panels on the sidebar "Typography" and "Text Settings", I believe that's essentially because the remaining paragraph customizations (drop cap, font-sizes) might need to be refactored to rely on the CSS vars implementation too. An alternative in the meantime is possible by introducing Slots (like we do for the advanced section) but maybe the double pannel is fine for now.

I'd appreciate some thoughts on the approach and if you like it, feel free to include in this PR.

ItsJonQ replied:
@youknowriad I'm in favour of this approach! It feels more inline with colors and global styles.

> Right now, the biggest downside is that we end up with two panels on the sidebar "Typography" and "Text Settings"

Yes... that's a bit of a concern 😅 

In theory, once more (style) attributes move over to this new system, it'll be smoothed out. This may be the first of many in that migration.

The only issue I can see is that the base style isn't as specific enough to override the styles from themes (or even the editor).

I'll merge your updates in https://github.com/WordPress/gutenberg/commit/6c7dcbaa9458659254c7d169cf791918bf0dd9eb into this PR

jorgefilipecosta commented on packages/block-editor/src/components/line-height-control/index.js:
```diff
@@ -0,0 +1,81 @@
+				// Increment by step value
+				adjustedNextValue = BASE_DEFAULT_VALUE + STEP;
+				break;
+			case '0':
```
If I write a value of 0 directly we assume right away another value as if the user was decreasing the value.
If I write a value of 0.1 directly we assume right away another value as if the user was increasing the value.

Maybe in the event that is triggered, we can check if the controls up/down are being used instead of using the value as a way to check?

ItsJonQ replied:
Oh, good call. This one is tricky. We can capture `UP`/`DOWN` on keyboard, but not so much clicking `up`/`down` in the `input` control.

Investigating...

ItsJonQ replied:
@jorgefilipecosta Haii! I just pushed a change to account for this logic.

Typing in `0` (zero) has now been accounted for.

Some good news is, the user can't physically trigger the `0.1` case by typing. This flow can only be triggered by either pressing `UP` or  clicking the `up` arrow in the `input`.

youknowriad committed:
Use CSS variables and a support flag for the line height customization

ZebulanStanphill commented on packages/block-editor/src/components/line-height-control/utils.js:
```diff
@@ -0,0 +1,49 @@
+ * @return {boolean} Whether the lineHeight attribute is valid.
+ */
+export function isLineHeightDefined( lineHeight ) {
+	return ! isUndefined( lineHeight ) && lineHeight !== RESET_VALUE;
```
```suggestion
	return lineHeight !== undefined && lineHeight !== RESET_VALUE;
```

ItsJonQ replied:
Thanks for the suggestion @ZebulanStanphill ! I'll commit it. I'll need to update the file too as well to remove the `import`.

Commit:
Improves handling of "zero" key press

Commit:
Update packages/block-editor/src/components/line-height-control/utils.js

Co-Authored-By: Zebulan Stanphill <zebulanstanphill@protonmail.com>

Commit:
Remove unused import

ItsJonQ commented:
@youknowriad 👋 Haiii!

The latest in Paragraph x LineHeight adventures.
The styles are no longer being applied because of the line-height specificity.

```css
p {
	line-height: var(--wp--typography--line-height);
}
```

Both the Editor and the front-end of the site have rules that override the `line-height` variable specified by the block.

Bumping up the specificity of `p` with something like:

`:root p` or `html p` may not be strong enough in most cases. However, these approaches are probably the most hands-off attempts in doing so.

In my initial update, I had added a `has-line-height` class to the `p` tag to help (a bit) with the specialty issue. I recognize, both then and now, that that approach isn't something that can scale well long term

---

**TLDR***

We need a... controlled (but agnostic) way, to ensure better consistency with CSS variable rendering. This issue isn't new 😅 . But I think we're able to identify opportunities and strategies now due to initiatives like Global Styling.

For **this PR**... What can we do?

We could leverage the `has-line-height` className based mechanism for now. As that is how most of Gutenberg currently works.

Would love some thoughts

Thank you all! <3

youknowriad commented:
This is in general coming from theme styles. It might work in some themes but not others, I wonder if documentation and communication is the solution here.

ItsJonQ commented:
> This is in general coming from theme styles. It might work in some themes but not others, I wonder if documentation and communication is the solution here.

@youknowriad Ultimately, I think documentation and communication will be the solution for the way we're handling styles.

For example, this update (as it is right now), won't work TwentyTwenty, due to it's custom `p` specificity 😅

youknowriad commented:
After some discussions in #21037 We decided to move forward with `:root p` specificity for now which will make this work in a lot of themes but not all of them (2019 works but not 2020), we'll be writing about this for the next Gutenberg plugin to let theme authors know how they can adapt to this change and we'll be monitoring feedback.

youknowriad commented:
I rebased this branch, for me aside potential test failures (wait for travis), this is ready.

ItsJonQ commented:
Thank you so much @youknowriad !! 🤗 
And to everyone for your amazing feedback and testing!

youknowriad committed:
Remove useless hooks file

ItsJonQ commented:
Oh boy!! Looks like Travis is green 🙌 

Will merge it up!

mwargan commented:
@ItsJonQ when disabling both font sizes and line height, I still see the Typography panel. Shouldn't this be removed?

ItsJonQ commented:
@mwargan Ah yes! I submitted a PR here that fixes that case:
https://github.com/WordPress/gutenberg/pull/22605

afercia commented:
Late to the party. While I do realize there's sometimes the need to set custom line-heights, I also do share some of @ZebulanStanphill's concerns.

**General considerations:**
1. Putting all this power in the users hands is potentially dangerous. Ensuring "true WYSIWYG" is one of the block editor goals but it shouldn't allow limitless customization power. There should be a limit to what users can and cannot do. Allowing users to indiscriminately play with font-sizes and line-heights potentially exposes to a complete disruption of the theme typography, potentially resulting in arguable results that will break readability of the front end content.
2. The block editor should establish limits to the allowed line-height values. Too small values shouldn't be allowed as they harm readability. 

**Architectural considerations:**
1. The current implementation provides the ability to set only _custom_ values. As such, the line-height value is saved as an inline style and permanently stored in the HTML blob in the database. This breaks the principle of separation of content from presentation. It also breaks interoperability of the data: exporting the content for reuse in other applications will bring all these presentational values in the content, where they would likely break the presentation. This brings us back to 15 years ago when web content was often an inextricable "HTML soup".
2. Admittedly, this is not new to the line-height control. It already happens for _custom_ font sizes etc. But, _instead of moving away from inline styles, I see the block editor is keeping using inline styles_. Maybe a bit too lightly. Instead of _more_ inline styles, I'd expect Gutenberg to explore new solutions to make _custom_ presentational values be stored separately from content. Note that the _preset_ values use CSS classes insead, which is way better.
3. As of July 2020, WordPress is powering 37.8% of the top 10 million websites in the world. This means the _block editor is being polluting the content of 37.8% of the websites with presentational data_. Not a great progress for the Web overall. I'd like the block editor team to be more aware of this and feel a stronger responsibility to make the Web better.

**Implementation considerations:**
1. At the very least, id' agree the line-height control should be paired with the font-size control and provide _preset_ values in addition to the custom one. This would also encourage to use the preset values (CSS classes) instead of custom value (inline style).
2. A line-height value makes truly sense only in relationship with the font-size is used for. Only themes are ultimately responsible for the content font-size.
3. When themes provide support for `editor-font-sizes` the should be required to pair a set of line-height values for each present font-size. Only this way, the line-height value makes sense because it's crafted on a specific font-size value. This would help consistency in the design and mitigate the "HTML soup" problem. Users can always set a custom value if they'r enot satisfied with the presets one.
4. When themes do _not_ provide support for `editor-font-sizes`, the block editor should provide sets of preset line-height values for each defaults in the `fontSizes` array. Again, only this way line-height values can be well-crafted in relationships to the actual font-size value.

**Current user interface shortcomings:**
I will create a separate, more detailed, issue. Also, in combination with some changes to the `editor-styles-wrapper` the preview in the block editor is not accurate. Some quick considerations (will better detail in the separate issue):
1. Some themes use a default line-height with several decimal values. For example, Twenty Twenty  uses `1.476` for the content paragraphs. The editor UI doesn't allow to set this value other than manually. The `0.1` step value for the `number` input type doesn't seem appropriate. I wonder whether a `number` input is appropriate in the first place but I can't think of of a better control off the top of my head.
2. There's no clear way to _reset_ a custom line-height value. The Backspace key works but that's hardly discoverable. At the very least, there should be a "Reset" button like for other custom values inputs.
3. The placeholder value of `1.5` is highly confusing. At first sight, it seems the default value, while the actual value from the theme may be completely different. Not sure there's a reason to display the placeholder in the first place and I'd tend to think it should be removed. Noting that the font-size _custom_ value doesn't use a placeholder, for good reasons.

<img width="365" alt="line height control" src="https://user-images.githubusercontent.com/1682452/87249382-174bc400-c45f-11ea-8f23-61bb3a719877.png">

kryslynn230 reviewed (approved):
Approved
